metricflow/inference/solver/weighted_tree.py

```diff
@@ -0,0 +1,163 @@
+        confidence_weight_map = {
+            InferenceSignalConfidence.LOW: 1,
+            InferenceSignalConfidence.MEDIUM: 2,
+            InferenceSignalConfidence.HIGH: 4,
+            InferenceSignalConfidence.FOR_SURE: 8,
+        }
+
+        return confidence_weight_map[confidence]
+
+    def __init__(
+        self, weight_percent_threshold: float = 0.6, weighter_function: Optional[NodeWeighterFunction] = None
+    ) -> None:
+        """Initialize the solver.
+
+        weight_percent_threshold: a number between 0.5 and 1. If a node's weight corresponds to a percentage
+            above this threshold with respect to its siblings' total weight sum, the solver will progress deeper
+            into the type tree, entering that node. If not, it stops at the parent.
+        weighter_function: a function that returns a weight given a confidence score. It will be used
+            to assign integer weights to each node in the type tree based on its input signals.
+        """
+        assert weight_percent_threshold >= 0.5 and weight_percent_threshold <= 1
+        self._weight_percent_threshold = weight_percent_threshold
+
+        self._weighter_function = (
+            weighter_function
+            if weighter_function is not None
+            else WeightedTypeTreeInferenceSolver.default_weighter_function
+        )
+
+    def _get_cumulative_weights_for_root(
+        self,
+        root: InferenceSignalNode,
+        weights: Dict[InferenceSignalNode, int],
+        non_complimentary_weights: Dict[InferenceSignalNode, int],
+        signals_by_type: Dict[InferenceSignalNode, List[InferenceSignal]],
+    ) -> Dict[InferenceSignalNode, int]:
+        """Get a dict of cumulative weights, starting at `root`.
+
+        A parent node's weight is the sum of all its children plus its own weight. Complimentary children get
+        excluded from the parent's sum.
+
+        root: the root to start assigning cumulative weights from.
+        weights: the output dictionary to assign the cumulative weights to
+        non_complimentary_weights: similar to weights, but the weight of each node excludes the weights
+            of all of its complimentary children
+        signals_by_type: a dictionary that maps signal type nodes to signals
+        """
+        for child in root.children:
+            self._get_cumulative_weights_for_root(
+                root=child,
+                weights=weights,
+                non_complimentary_weights=non_complimentary_weights,
+                signals_by_type=signals_by_type,
+            )
+
+        weights[root] = sum(self._weighter_function(signal.confidence) for signal in signals_by_type[root])
+        non_complimentary_weights[root] = sum(
+            self._weighter_function(signal.confidence)
+            for signal in signals_by_type[root]
+            if not signal.is_complimentary
+        )
+
+        weights[root] += sum(non_complimentary_weights[child] for child in root.children)
```

Comment:
These aren't order dependent, right? If not, swapping like this (with renames as needed) would be easier to understand:
```suggestion
        weights[root] = sum(self._weighter_function(signal.confidence) for signal in signals_by_type[root])
        weights[root] += sum(non_complimentary_weights[child] for child in root.children)

        non_complimentary_weights[root] = sum(
            self._weighter_function(signal.confidence)
            for signal in signals_by_type[root]
            if not signal.is_complimentary
        )
```